### PR TITLE
IOS-17850: cache the map of background session ID and its associateID…

### PR DESCRIPTION
**Pull request by thuyBox**
…s under backgroundSessionsIndex

### CODE REVIEW

**Comment by boxcla** [comment]
Verified that @thuyBox has signed the CLA. Thanks for the pull request!

File: BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m [hunk 1]
```diff
@@ -850,30 +890,27 @@ - (NSString *)filePathOfExtensionBackgroundSessionId:(NSString *)backgroundSessi
...
-    for (NSString *associateId in associateIdToBackgroundSessionIdAndSessionTaskIdDict) {
-        BOXURLBackgroundSessionIdAndSessionTaskId *backgroundSessionIdAndSessionTaskId = associateIdToBackgroundSessionIdAndSessionTaskIdDict[associateId];
+    BOOL isDir = NO;
+    NSMutableArray *associateIDs = nil;
 
-        if ([backgroundSessionIdAndSessionTaskId.backgroundSessionId isEqualToString:backgroundSessionId]) {
-            [associateIds addObject:associateId];
+    if ([[NSFileManager defaultManager] fileExistsAtPath:dir isDirectory:&isDir] == YES && isDir == YES) {
+        NSArray *filePaths = [[NSFileManager defaultManager] contentsOfDirectoryAtPath:dir error:error];
+        if (filePaths.count > 0) {
```
Comment (jlawton) [inline BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m:905]: I think we should follow the system conventions, and only return `nil` on an error (`filePaths == nil`).

File: BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.h [hunk 2]
```diff
@@ -280,6 +283,17 @@
...
 
+/**
+ * Check if a background session is valid, i.e. exists for a userId
+ *
+ * @param userID            Id of user
+ * @param backgroundSessionID       Id of background session
+ *
+ * @return YES if valid, NO if not
+ */
+- (BOOL)isBackgroundSessionValidGivenUserID:(NSString * _Nonnull)userID
```
Comment (jlawton) [inline BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.h:294]: I think that `Given` is less natural than `For` in these method names.

For this specific name, the idea of `valid` is a little strange. Maybe something like `isBackgroundSessionID:associatedWithUserID:` would be clearer?

**Review by jlawton — APPROVED** [review]

File: BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m (hunk 1 above)
Comment (thuyBox) [reply BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m:905]: updated in follow-up commit

File: BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.h (hunk 2 above)
Comment (thuyBox) [reply BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.h:294]: updated in follow-up commit

**Pull request merged**